ttnn/cpp/ttnn/tensor/types.cpp

```diff
@@ -293,3 +293,28 @@ MemoryConfig load_memory_config(const std::string& file_name) {
 }  // namespace tt_metal
 
 }  // namespace tt
+
+namespace ttnn {
+namespace types {
+
+uint32_t &Shape::operator[](const std::int64_t index) {
+    auto normalized_index = this->get_normalized_index(index);
+    return this->value[normalized_index];
+}
+
+const uint32_t Shape::operator[](std::int64_t index) const { return this->value.without_padding()[index]; }
+
+const uint32_t* Shape::begin() const {return this->value.begin(); }
+const uint32_t* Shape::end() const {return this-> value.end(); }
+
+tt::tt_metal::Array4D Shape::to_array_4D() const {
+    tt::tt_metal::Array4D ret_array;
+    for (int i = 0; i < rank(); i++) {
```

Comment:
no guarantee that its rank is 4D

Reply:
This one we can probably just remove. Use is limited and we can fix in usage.

Reply:
or i can add an assert to check if it's 4D?

Reply:
Adding as a check for now